src/whir/batching.rs

```diff
@@ -0,0 +1,819 @@
+///   oracles.)
+///
+/// - During proof verification, for the first round, the Verifier validates
+///   each individual STIR query in the proof against individual Merkle root.
+///   Once these Merkle paths are validated, the Verifier re-derives
+///   batching_randomness and combines the STIR responses using powers of
+///   batching_randomness.
+///
+/// - After the first round, rest of the protocol proceeds as usual.
+///
+use ark_crypto_primitives::merkle_tree::{
+    Config as MTConfig, LeafParam, MerkleTree, MultiPath, TwoToOneParam,
+};
+use ark_ff::{FftField, Field};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+#[cfg(any(feature = "parallel", feature = "rayon"))]
+use rayon::prelude::*;
+use spongefish::{
+    codecs::arkworks_algebra::{
+        ByteDomainSeparator, BytesToUnitDeserialize, BytesToUnitSerialize, FieldDomainSeparator,
+        FieldToUnitDeserialize, FieldToUnitSerialize, UnitToField,
+    },
+    ProofError, ProofResult, UnitToBytes,
+};
+use spongefish_pow::PoWChallenge;
+
+use crate::{
+    fs_utils::{OODDomainSeparator, WhirPoWDomainSeparator},
+    poly_utils::coeffs::CoefficientList,
+    sumcheck::SumcheckSingleDomainSeparator,
+    whir::{
+        committer::{CommitmentReader, CommitmentWriter, Witness},
+        domainsep::{DigestDomainSeparator, WhirDomainSeparator},
+        parameters::WhirConfig,
+        prover::Prover,
+        statement::{Statement, StatementVerifier},
+        utils::{DigestToUnitDeserialize, DigestToUnitSerialize},
+        verifier::Verifier,
+        RoundZeroProofValidator, WhirCommitmentData, WhirProof, WhirProofRoundData,
+    },
+};
+
+fn fma_stir_queries<F: Field>(
+    scale_factor: F,
+    stir_queries: &[Vec<F>],
+    folded_queries: &mut [Vec<F>],
+) {
+    for (folded_vec, stir_vec) in folded_queries.iter_mut().zip(stir_queries.iter()) {
+        for (i, value) in stir_vec.iter().enumerate() {
+            folded_vec[i] += scale_factor * value
+        }
+    }
+}
+
+#[cfg(debug_assertions)]
+fn validate_stir_queries<F, MerkleConfig>(
+    batching_randomness: &F,
+    expected_stir_value: &[Vec<F>],
+    stir_list: &[(MultiPath<MerkleConfig>, Vec<Vec<F>>)],
+) -> bool
+where
+    F: Field,
+    MerkleConfig: MTConfig<Leaf = [F]>,
+{
+    if stir_list.is_empty() {
+        return true;
+    }
+
+    let (basic_indexes, mut computed_stir) = stir_list[0].clone();
+    let mut multiplier = *batching_randomness;
+
+    for (multi_path, stir_queries) in &stir_list[1..] {
+        if !multi_path
+            .leaf_indexes
+            .iter()
+            .zip(basic_indexes.leaf_indexes.iter())
+            .all(|(lhs, rhs)| lhs == rhs)
+        {
+            return false;
+        }
+
+        fma_stir_queries(multiplier, stir_queries.as_slice(), &mut computed_stir);
+        multiplier *= batching_randomness;
+    }
+
+    let result = expected_stir_value == computed_stir;
+    result
+}
+
+#[cfg(not(debug_assertions))]
```

Comment:
I would apply the `#[cfg(debug_assertions)]` on the call site. Or use `debug_assert!` which has this built in.